pkgs/racket-doc/scribblings/reference/eval-model.scrbl

```diff
@@ -60,55 +59,57 @@ Some simplifications require more than one step. For example:
 ]
 
 An expression that is not a @tech{value} can always be partitioned
-into two parts: a @deftech{redex}, which is the part that changed in a
+into two parts: a @deftech{redex} (``reducible expression''),
+which is the part that can change in a
 single-step simplification (highlighted), and the
 @deftech{continuation}, which is the evaluation
-context surrounding an expression. In @racket[(- 4 (+ 1 1))], the redex is @racket[(+ 1 1)], and
-the continuation is @racket[(- 4 @#,hole)], where @hole takes the
-place of the redex. That is, the continuation says how to ``continue''
+context surrounding the redex. In @racket[(- 4 (+ 1 1))], the redex is @racket[(+ 1 1)], and
+the continuation is @racket[(- 4 @#,hole)], where @hole indicates the
+position of the redex. That is, the continuation says how to ``continue''
 after the @tech{redex} is reduced to a @tech{value}.
 
-Before some things can be evaluated, some sub-expressions must be
-evaluated; for example, in the application @racket[(- 4 (+ 1 1))], the
+Before some expressions can be evaluated, their sub-expressions must be
+evaluated. For example, in the application @racket[(- 4 (+ 1 1))], the
 application of @racket[-] cannot be reduced until the sub-expression
 @racket[(+ 1 1)] is reduced.
-
 Thus, the specification of each syntactic form specifies how (some of)
-its sub-expressions are evaluated, and then how the results are
+its sub-expressions are evaluated and then how the results are
 combined to reduce the form away.
 
 The @deftech{dynamic extent} of an expression is the sequence of
-evaluation steps during which the expression contains the @tech{redex}.
+evaluation steps starting when the expression is selected as a @tech{redex}
+and continuing until it is fully evaluated.
```

Comment:
I think this change is incorrect. The original wording implies that the dynamic extent of an expression is the sequence of reduction steps when the expression *contains* the redex, that is, contains the hole. This means that in the evaluation of the expression `(+ 1 (+ 2 3))`, the dynamic extent of the expression includes the step of evaluation that decomposes the expression into the context `(+ 1 [])` and the redex `(+ 2 3)`. However, the new wording would exclude that step, since the expression itself is not the redex.

Reply:
I have reverted to the original text, but I still do not understand.

In your second sentence above, "contains the redex" seems ambiguous, since an expression can contain multiple redexes.

Reply:
No, it can’t. While I’m not sure it is explicitly explained in the current reference, there are rules that ensure each expression has a *unique decomposition* into a continuation and a redex. For example, in the expression `(+ (* 2 3) (* 4 5))`, *the* redex is `(* 2 3)`, since Racket defines a left-to-right evaluation order. The `(* 4 5)` expression can only become the redex once `(* 2 3)` has been reduced to a value, and the overall expression becomes `(+ 6 (* 4 5))`.

Reply:
> So we don't say that the dynamic extent of the outer expression includes both reductions?

We do! But note that “dynamic extent” is just defined as “the sequence of **evaluation steps**” (emphasis mine) during with the expression contains the redex. Since the decomposition (into continuation+redex), reduction (of the redex), and recomposition (of the continuation+reduced redex) process comprises a *single* evaluation step, it doesn’t make sense to think of dynamic extent of applying to some of those “sub-steps” but not others.

Given the above, and given that each evaluation step reduces exactly one redex one step, by definition, the phrase “the sequence of evaluation steps during which the expression contains the redex” is precise. There is only one redex per evaluation step, and an expression contains the redex the unique decomposition selects a redex inside the expression.

Reply:
I'm going to keep going even though I'm starting to wonder whether it's just me.

I'm confused about whether a dynamic extent includes only one redex. Above you said that it cannot, but then in your post just now you said that the dynamic extent of the outer expression includes both reductions [the expression `(+ (* 2 3) (* 4 5))`]. So does the dynamic extent of that expression include the reduction of only `(* 2 3)` or does it include the reduction of both arguments?

If it is only `(* 2 3)`, then wouldn't the following be clearer: "... contains the redex or its hole." After all, we are emphasizing that the hole really does take the place of the redex during reduction.

On the other hand, if it is both the subexpressions, then it doesn't seem right to refer to "_the_ redex."

Reply:
> I'm confused about whether a dynamic extent includes only one redex. Above you said that it cannot, but then in your post just now you said that the dynamic extent of the outer expression includes both reductions

I never said that dynamic extent can’t have several redexes, merely that each expression has a unique decomposition into a continuation and a redex. There is one redex per evaluation step, by definition.

I think I can see now where your confusion is coming from, though: the reference plays a little loose with what “an expression” is. It’s true that `(+ (* 2 3) (* 4 5))` has exactly one redex—`(* 2 3)`—but after a single evaluation step, it becomes `(+ 6 (* 4 5))`, which has a new redex—`(* 4 5)`. The surrounding `(+ ....)` has stuck around, but the expression is fundamentally a new expression… it must be for some evaluation to have taken place!

In that sense, in order to fully specify dynamic extent, we have to think not just about the redex but also about the continuation. In both evaluation steps, the application expression `(+ ....)` is in the continuation, unchanging, and it contains the hole (and therefore the redex). We say that this is a single expression with a notion of continuity over several evaluation steps, even though its subexpressions are being reduced.

Perhaps someone else can supply a better way of describing this. The current wording matches my intuition well, so I am perhaps not the best person to suggest an improvement.

Reply:
I don't see how `(+ ....)` can contain only one hole. The first redex is `(* 2 3)`, which is replaced by a hole. But then the expression is reduced to `(+ 6 (* 4 5))`, which has a second redex.

If that second expression is considered to be the original expression, then the original expression has two redexes.

If that second expression is considered to be a new expression, then the dynamic extent of the original expression doesn't cover the evaluation of `(* 4 5)`, because it is finished "containing a redex."

Thanks for your patience with this. We are making progress!